JWBestPracticeApps/Casting/CastingViewController.swift

```diff
@@ -0,0 +1,182 @@
+        
+        self.castingButton = castingButton
+        self.barButtonItem = barButtonItem
+        
+        setCastingStatus(false)
+    }
+    
+    func prepareCastingButtonWithAnimation(_ button: UIButton) {
+        let connectingImages = [UIImage(named: "cast_connecting0")?.withRenderingMode(.alwaysTemplate),
+                                UIImage(named: "cast_connecting1")?.withRenderingMode(.alwaysTemplate),
+                                UIImage(named: "cast_connecting2")?.withRenderingMode(.alwaysTemplate),
+                                UIImage(named: "cast_connecting1")?.withRenderingMode(.alwaysTemplate)]
+        // Compact map to avoid nil UIImage objects
+        button.imageView?.animationImages = connectingImages.compactMap {$0}
+        button.imageView?.animationDuration = 2
+    }
+    
+    
+// pragma MARK: - Casting Status Helpers
+    
+    func startConnectingAnimation() {
+        castingButton?.tintColor = UIColor.white
+        castingButton?.imageView?.startAnimating()
+    }
+    
+    func stopConnectingAnimation(connected: Bool) {
+        castingButton?.imageView?.stopAnimating()
+        let castingImage = connected ? "cast_on" : "cast_off"
+        castingButton?.setImage(UIImage(named: castingImage)?.withRenderingMode(.alwaysTemplate), for: .normal)
+        castingButton?.tintColor = UIColor.blue
+    }
+    
+    func setCastingStatus(_ casting: Bool) {
+        self.casting = casting
+        castingButton?.tintColor = casting ? UIColor.green : UIColor.blue
+    }
+    
+    @objc func castButtonTapped(sender: UIButton) {
+        let alertController = UIAlertController(title: nil, message: nil, preferredStyle: .actionSheet)
+        if let device = castController?.connectedDevice {
+            alertController.title = device.name
+            alertController.message = "Select an action"
+            
+            let disconnetAction = UIAlertAction(title: "Disconnect", style: .destructive) { [weak self] (action) in
+                guard let self = self else { return }
+                self.castController?.disconnect()
+            }
+            
+            if self.casting {
+                alertController.addAction(UIAlertAction(title: "Stop casting", style: .default, handler: { [weak self] (action) in
+                    guard let self = self else { return }
+                    self.castController?.stopCasting()
+                }))
+            } else {
+                alertController.addAction(UIAlertAction(title: "Cast", style: .default, handler: { [weak self] (action) in
+                    guard let self = self else { return }
+                    self.castController?.cast()
+                }))
+            }
+            alertController.addAction(disconnetAction)
+        } else {
+            alertController.title = "Connect to"
+            self.castController?.availableDevices.forEach({ (castingDevice) in
```

Comment:
I'm not sure about this. I think since `forEach` has a non-escaping closure I don't need to capture `self` inside its closure, but I put a `[weak self]` in the next line because the UIAlertAction closure need to be executed asynchronous. @henryhlee what do you think?